### PR TITLE
[serve] Catch `RuntimeEnvSetupError` when checking status

### LINKED ISSUE #26205: [serve] Bad runtime_env causes traceback instead of a status message
Deployed a config with a bad runtime_env, e.g., `runtime_env: {"fake": "news"}`

```
ray.exceptions.RuntimeEnvSetupError: Failed to setup runtime environment.
ray::ServeController.get_serve_status() (pid=40678, ip=127.0.0.1, repr=<ray.serve.controller.ServeController object at 0x10db66520>)
  File "/Users/eoakes/code/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 360, in _create_runtime_env_with_retry
    runtime_env_context = await asyncio.wait_for(
  File "/Users/eoakes/miniforge3/envs/ray/lib/python3.8/asyncio/tasks.py", line 494, in wait_for
    return fut.result()
  File "/Users/eoakes/code/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 317, in _setup_runtime_env
    await loop.run_in_executor(None, setup_plugins)
  File "/Users/eoakes/miniforge3/envs/ray/lib/python3.8/concurrent/futures/thread.py", line 57, in run
    result = self.fn(*self.args, **self.kwargs)
  File "/Users/eoakes/code/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 303, in setup_plugins
    raise RuntimeError(f"runtime env plugin {name} not found.")
RuntimeError: runtime env plugin fake not found.
```

### PR DESCRIPTION
**Pull request by zcin**
Signed-off-by: Cindy Zhang <cindyzyx9@gmail.com>

## Why are these changes needed?

When the task `run_graph` is submitted, fetching the object ref result can raise `RayTaskError` if the task fails or `RuntimeEnvSetupError` if setting up the runtime env for the task fails. We want to deal with the latter case as well.

## Related issue number

"Closes #26205"

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by zcin** [comment]
Windows test unrelated.

**Review by sihanwang41 — APPROVED** [review]
Nice!

**Review by shrekris-anyscale — APPROVED** [review]
Great catch! Do you mind pasting a sample traceback when the `runtime_env` setup fails? I'm curious what it looks like.

**Comment by zcin** [comment]
> Great catch! Do you mind pasting a sample traceback when the `runtime_env` setup fails? I'm curious what it looks like.

Yup, an example is
```
    Traceback (most recent call last):
      File "/Users/cindyz/ray/python/ray/serve/controller.py", line 582, in get_serve_status
        await finished[0]
    ray.exceptions.RuntimeEnvSetupError: Failed to set up runtime environment.
    Traceback (most recent call last):
      File "/Users/cindyz/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 342, in _create_runtime_env_with_retry
        runtime_env_context = await asyncio.wait_for(
      File "/Users/cindyz/miniforge3/envs/ray/lib/python3.8/asyncio/tasks.py", line 494, in wait_for
        return fut.result()
      File "/Users/cindyz/ray/python/ray/dashboard/modules/runtime_env/runtime_env_agent.py", line 297, in _setup_runtime_env
        await create_for_plugin_if_needed(
      File "/Users/cindyz/ray/python/ray/_private/runtime_env/plugin.py", line 252, in create_for_plugin_if_needed
        size_bytes = await plugin.create(uri, runtime_env, context, logger=logger)
      File "/Users/cindyz/ray/python/ray/_private/runtime_env/working_dir.py", line 155, in create
        local_dir = await download_and_unpack_package(
      File "/Users/cindyz/ray/python/ray/_private/runtime_env/packaging.py", line 686, in download_and_unpack_package
        with open_file(pkg_uri, "rb", transport_params=tp) as package_zip:
      File "/Users/cindyz/miniforge3/envs/ray/lib/python3.8/site-packages/smart_open/smart_open_lib.py", line 224, in open
        binary = _open_binary_stream(uri, binary_mode, transport_params)
      File "/Users/cindyz/miniforge3/envs/ray/lib/python3.8/site-packages/smart_open/smart_open_lib.py", line 400, in _open_binary_stream
        fobj = submodule.open_uri(uri, mode, transport_params)
      File "/Users/cindyz/miniforge3/envs/ray/lib/python3.8/site-packages/smart_open/s3.py", line 221, in open_uri
        parsed_uri = parse_uri(uri)
      File "/Users/cindyz/miniforge3/envs/ray/lib/python3.8/site-packages/smart_open/s3.py", line 112, in parse_uri
        head, key_id = uri.split('/', 1)
    ValueError: not enough values to unpack (expected 2, got 1)
```

File: python/ray/serve/controller.py [hunk 1]
```diff
@@ -583,6 +583,11 @@ async def get_serve_status(self) -> bytes:
                 except RayTaskError:
                     serve_app_status = ApplicationStatus.DEPLOY_FAILED
                     serve_app_message = f"Deployment failed:\n{traceback.format_exc()}"
+                except RuntimeEnvSetupError:
+                    serve_app_status = ApplicationStatus.DEPLOY_FAILED
+                    serve_app_message = (
+                        f"Runtime env setup failed:\n{traceback.format_exc()}"
+                    )
```
Comment (edoakes) [inline python/ray/serve/controller.py:591]: I would suggest we have a generic `except Exception` block here that also makes it `DEPLOY_FAILED` with a message something along the lines of "unknown error occurred: {exception_message}"

**Review by edoakes — COMMENTED** [review]
It's strange that `RuntimeEnvSetupError` doesn't inherit from `RayTaskError`, seems like a footgun 🤔 we might want to consider proposing that change on the core side.

cc @rkooo567, I think he implemented it originally

File: python/ray/serve/tests/test_cli.py [hunk 2]
```diff
@@ -272,6 +272,26 @@ def check_for_failed_deployment():
...
+        os.path.dirname(__file__), "test_config_files", "bad_runtime_env.yaml"
+    )
+
+    subprocess.check_output(["serve", "deploy", config_file_name])
+
+    def check_for_failed_deployment():
+        app_status = ServeSubmissionClient("http://localhost:52365").get_status()
+        return app_status["app_status"]["status"] == "DEPLOY_FAILED"
+
+    wait_for_condition(check_for_failed_deployment)
```
Comment (architkulkarni) [inline python/ray/serve/tests/test_cli.py:292]: I'd suggest also checking somewhere in the test that the user-friendly runtime env error output (e.g. s3 bucket not found, or pip package does not exist) appears in the expected place.  (You can just check for existence of a relevant substring)

**Review by architkulkarni — APPROVED** [review]

**Review by edoakes — APPROVED** [review]

**Comment by edoakes** [comment]
@zcin a runtime_env test was failing on CI which might be related. I think it isn't and was the issue that @architkulkarni fixed in a recent PR, but can you re-merge master and push to trigger another build?

**Comment by zcin** [comment]
Tests look good! Windows test `test_numba` still failing on master.

**Pull request merged**